### PR TITLE
argument completion of R method when pipe

### PR DESCRIPTION
**Pull request by romainfrancois**
### Intent

addresses #11627

### Approach

special case based on argument `chainObjectName`

### Automated Tests

> Indicate whether this change includes unit tests or integration tests, or link a work item or pull request to add those tests to another repo. If the change cannot or will not be covered by a test, indicate why.

### QA Notes

> Add additional information for QA on how to validate the change, paying special attention to the level of risk, adjacent areas that could be affected by the change, and any important contextual information not present in the linked issues. 

### Checklist

- [ ] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md` 
- [ ] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [ ] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [ ] This PR passes all local unit tests

### CODE REVIEW

**Comment by romainfrancois** [comment]
(copied from initial pr): 

I need to study some more what happens here, and if we ever get there: 

https://github.com/rstudio/rstudio/blob/710e8ec5601997f8e10bcc57126c5fb0b090ebf2/src/cpp/session/modules/SessionRCompletions.R#L2115-L2145

But what this pr does is inject the "chain object name" e.g. `mtcars` into the `functionCallString`, so that `.rs.getRCompletions()` can take advantage of it. 

image

BTW, should column names (mpg, cyl, etc ...) look a bit different in the completion popup besides not having the `=` ? Maybe a different icon, or font weight ?

> I need to study some more what happens here, and if we ever get there:
> 
> https://github.com/rstudio/rstudio/blob/710e8ec5601997f8e10bcc57126c5fb0b090ebf2/src/cpp/session/modules/SessionRCompletions.R#L2115-L2145
> 
> But what this pr does is inject the "chain object name" e.g. `mtcars` into the `functionCallString`, so that `.rs.getRCompletions()` can take advantage of it.
> 
> image
> 
> BTW, should column names (mpg, cyl, etc ...) look a bit different in the completion popup besides not having the `=` ? Maybe a different icon, or font weight ?

I think different icon make more sense to me @romainfrancois 


It'd also be nice if the column names were displayed first (rather than last) in the completion list. I also like the idea of including the origin of the matches.

Maybe this should happen in `.rs.sortCompletions()`?

https://github.com/rstudio/rstudio/blob/aa90ba6df9e8a8647c8ca007ba59447b6401615e/src/cpp/session/modules/SessionRCompletions.R#L1214-L1236

It would also need to handle the case where "token" is empty.

https://github.com/rstudio/rstudio/blob/aa90ba6df9e8a8647c8ca007ba59447b6401615e/src/cpp/session/modules/SessionRCompletions.R#L2571-L2572

Currently at: 

image

I have not figured out yet how to sort the columns before the arguments. 

Looks great so far!

> I have not figured out yet how to sort the columns before the arguments.

You might also need to look around here:

https://github.com/rstudio/rstudio/blob/79f339a38c19b1324adb035504b05670cc0d61b7/src/gwt/src/org/rstudio/studio/client/workbench/views/console/shell/assist/CompletionRequester.java#L151-L217

(and perhaps other routines in that file)

For some feedback on the UI:

- For column name completions, can we also try a treatment where we use the "regular" data.frame icon, but with one column of the table shaded? (Just because I think a full-width icon may fit better; the single-column treatment feels too narrow)
- For the table name, I think I prefer the original treatment with the table name as `[mtcars]`. Do you have a strong opinion?

**Comment by romainfrancois** [comment]
revert to use `[mtcars]` and updated the icon to emphasize one column of the data frame. 

image

**Comment by romainfrancois** [comment]
image

... but I think argument first was better, because injecting columns won't make sense with every chain. I think we should revert the last 2 commits. 

Also Showing `... = ` is not that useful. Maybe we could condition showing the columns based on the presence of a `...` or perhaps functions like `mutate()` could have a way to declare this. 

For example, `dplyr::mutate()` has these: 

```r
#' @param ... <[`data-masking`][dplyr_data_masking]> Name-value pairs.
#'   The name gives the name of the column in the output.
#'
#'   The value can be:
#'
#'   * A vector of length 1, which will be recycled to the correct length.
#'   * A vector the same length as the current group (or the whole data frame
#'     if ungrouped).
#'   * `NULL`, to remove the column.
#'   * A data frame or tibble, to create multiple columns in the output.
```

Perhaps it could also have an attribute that would inform about how to handle `...`, and we could use this to surface some information in the completion.

**Comment by romainfrancois** [comment]
I believe this PR addresses the original issue in #11627, i.e. we get extra arguments in both piped and non piped cases: 

image

image

Although the non piped still shows the first argument `.data`.

**Comment by romainfrancois** [comment]
updated: 

image

image

I think it's good to go. @kevinushey 

 - I'll deal with the position of the help preview separately
 - I believe column completions should be done differently, perhaps based on some information the function makes available. I'll open an issue about this.

File: src/cpp/session/modules/SessionRCompletions.R [hunk 1]
```diff
@@ -917,27 +917,23 @@ assign(x = ".rs.acCompletionTypes",
       else
          ""
       
+      # arguments that are already used by the matched call
+      used <- names(as.list(matchedCall)[-1]) 
+      keep <- ! names(formals$formals) %in% used
```
Comment (kevinushey) [inline src/cpp/session/modules/SessionRCompletions.R:922]:
```suggestion
      keep <- !names(formals$formals) %in% used
```

File: src/cpp/session/modules/SessionRCompletions.R [hunk 2]
```diff
@@ -2141,6 +2137,15 @@ assign(x = ".rs.acCompletionTypes",
             functionCallString,
             fixed = TRUE
          )
+      } else if (nzchar(chainObjectName)) {
+         dropFirstArgument <- TRUE
```
Comment (kevinushey) [inline src/cpp/session/modules/SessionRCompletions.R:2141]: Looks like this isn't used anymore?

```suggestion
```

File: src/cpp/session/modules/SessionRCompletions.R [hunk 3]
```diff
@@ -2693,22 +2698,30 @@ assign(x = ".rs.acCompletionTypes",
...
-      {
+      if (length(object)) {
          objectNames <- .rs.getNames(object)
          if (length(objectNames))
          {
             completions <- .rs.selectFuzzyMatches(objectNames, token)
-            type <- vapply(completions, FUN.VALUE = numeric(1), USE.NAMES = FALSE, function(i) {
-               .rs.getCompletionType(object[[i]])
-            })
+            if (inherits(object, "data.frame")) {
```
Comment (kevinushey) [inline src/cpp/session/modules/SessionRCompletions.R:2706]:
```suggestion
            if (inherits(object, "data.frame"))
            {
```

**Comment by kevinushey** [comment]
One other thing you might want to take a look at -- in longer chains, we filter completions in a separate way. For example, code around here:

https://github.com/rstudio/rstudio/blob/8efe31a14b9881572cecef97d1f3089ca8a1a9b2/src/gwt/acesupport/acemode/r_code_model.js#L345-L373

The intention being that in code like this:

```
mtcars %>%
  select(mpg) %>%
  mutate(|)
```

we "know" the only column in scope is `mpg`, so we don't need to provide completions from `mpg` itself. Right now that's not getting your new UI treatment:

Screen Shot 2022-09-29 at 9 35 46 AM

**Comment by kevinushey** [comment]
(it's unfortunate that the completion system here is so divided across different components; in hindsight it would've been far better to consolidate as much of this as possible on the session side, but this seemed like the simplest way to tie things together at the time)

**Pull request merged**